include/rdma/fi_profile.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2023 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_PROFILE_H
+#define FI_PROFILE_H
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <rdma/fi_domain.h>
+
+/*
+ * pre-defined profiling variables
+ */
+enum {
+	FI_VAR_UNEXP_MSG_CNT,      // datatype: FI_UINT64
+	FI_VAR_UNEXP_MSG_QUEUE,    // datatype: FI_TYPE_CQ_ERR_ENTRY
+};
+
+/*
+ * pre-defined events which can change the profiling variables
+ */
+enum {
+	FI_EVENT_UNEXP_MSG_RECVD,    // var = FI_VAR_UNEXP_MSG_CNT
+	FI_EVENT_UNEXP_MSG_MATCHED,  // var = FI_VAR_UNEXP_MSG_CNT
+};
+
+struct fi_profile_desc {
+	uint32_t id;
+	union {
+		enum fi_datatype basic;
+		enum fi_type comp;
+	} type;
+	uint64_t flags;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+
+#define FI_VAR_TYPE_COMP  0x1   //  flags bit0,  0: basic, 1: comp
```

Comment:
This isn't easily extendable or intuitive.  Please switch to something more like this:
```
enum fi_profile_type {
    fi_primitive_type,
    fi_defined_type,
};

struct fi_profile_desc _
    ...
    enum fi_profile_type datatype;
    union {
        enum fi_datatype primitive;
        enum fi_type defined;
    } datatype;
    ...
```